doc/heps/hep2.md

```diff
@@ -0,0 +1,366 @@
+    - The `warn` function from the `warnings` module:
+
+    ```python
+    from warnings import warn
+
+    def foo():
+      warn(
+        "Function foo is deprecated since version 1.1.1 and will be removed in a future release, use bar instead.",
+        category=FutureWarning, stacklevel=2,
+      )
+      ...
+    ```
+
+    - The `deprecated` decorator added in Python 3.13 ([PEP 702](https://peps.python.org/pep-0702/)) to the `warnings` module, backported to previous Python versions via the [typing_extensions](https://github.com/python/typing_extensions).
+
+    ```python
+    from typing_extensions import deprecated
+
+    @deprecated(
+      "Function foo is deprecated since version 1.1.1 and will be removed in a future release, use bar instead.",
+      category=FutureWarning,
+    )
+    def foo():
+      ...
+    ```
+
+    The `deprecated` decorator not only emits a run-time warning, it also enables static type checkers to warn when they encounter usage of an object decorated with `deprecated`. This comes with various benefits, for example, VSCode users will see the deprecated objects crossed out in their code. Therefore, when possible, `deprecated` should be preferred over `warn`.
+
+  - The warning message:
+
+    - **Must** indicate that the feature is deprecated and is going to be removed in a future version.
+    - **Must** suggest replacement APIs, if applicable.
+    - Can indicate in which version the feature was deprecated.
+    - Can indicate before which version the feature is going to be remove, however, maintainers should ensure that the deprecation period (see below) is not going to be too to short, and that no obsolete warning will be released (e.g. feature announced to be removed in version 1.1 but is still present in 1.1).
+
+  - The warning type:
+  
+    - **Must** be a subclass of `DeprecationWarning`, `FutureWarning`, or `PendingDeprecationWarning`.
+    - **Must** be defined based on this approach (see the Appendix for more details):
+
+      1. Start by emitting a `DeprecationWarning`, to inform Library Developers, and some Data Analysts.
+      2. After a few months, upgrade to a `FutureWarning`, to inform all users.
+
+      Exceptions are allowed but should in practice be very rare:
+
+      - Start by emitting a `PendingDeprecationWarning` for deprecations that might be reverted or are expected to be extremely noisy.
+      - Don't upgrade to a `FutureWarning` when the feature removed is meant for Library Developers only.
+      - Emit directly a `FutureWarning` when the feature removed is meant for Data Analysts only.
+
+  - `stacklevel` **must** be set so the warning appears in the correct place. Projects can implement a utiliy like the private Pandas' function [`find_stack_level`](https://github.com/pandas-dev/pandas/blob/b8a4691647a8850d681409c5dd35a12726cd94a1/pandas/util/_exceptions.py#L34) to automatically infer the right `stacklevel` number.
```

Comment:
I don't see stacklevel being used in `@deprecated`

Reply:
Yep true, although it's part of its signature so there might be some reasons to use it for weird edge cases (can't think of one). I can update the text to reflect that. I also found out the runtime behavior of `@deprecated` can be disabled by setting `category` to `None` which means you can used both `@deprecated` and `warnings.warn` if needed. Which I think means that `typing_extensions` doesn't have to be added as a runtime dependency.

I also saw that `warnings.warn` will gain the new `skip_file_prefixes` parameter in Python 3.13. If I understand what it does I could mention it in the HEP.

Reply:
Pushed some updates in https://github.com/holoviz/holoviz/pull/388/commits/b5def36563835f171f95306b9eb9ee12384d0303 and https://github.com/holoviz/holoviz/pull/388/commits/f8e00677152545d4cc01938fd371b19fa7792f1e about the above.